traversal/planner/GraphPlanner.java

```diff
@@ -231,37 +233,43 @@ Optimiser optimiser() {
     }
 
     void mayOptimise(GraphManager graphMgr, boolean singleUse) {
-        if (procedure == null) {
-            try {
-                firstOptimisingLock.writeLock().lock();
-                if (procedure == null) optimise(graphMgr, singleUse);
-                assert procedure != null;
-            } finally {
-                firstOptimisingLock.writeLock().unlock();
-            }
-        } else if (isOptimising.compareAndSet(false, true)) {
-            try {
-                optimise(graphMgr, singleUse);
-            } finally {
-                isOptimising.set(false);
-            }
+        long timeLimitMillis = singleUse ? HIGHER_TIME_LIMIT_MILLIS : DEFAULT_TIME_LIMIT_MILLIS;
+        if (backgroundOptimisation == null) beginFirstOptimise(graphMgr, timeLimitMillis);
+        else if (isOptimising.compareAndSet(false, true)) beginReOptimise(graphMgr, timeLimitMillis);
+
+        try {
+            backgroundOptimisation.get(timeLimitMillis, MILLISECONDS);
```

Comment:
all threads wanting to use this plan give the background optimisation the required time to finish, and then return (this approach will immediately return if the optimisation is already done or as soon as it finishes, which is better than a `sleep()` that always consumes the maximum time)

Reply:
Unless the solver converges to an optimal solution in well under the allotted time, the optimisation will almost never finish within this time limit, because updating the costs and creating the procedure takes on the order of milliseconds. I'd suggest increasing the timeout here.

Reply:
from my profiling updating traversal costs can take around 2-5ms, and procedure creation a few more ms. I'll 10ms to the overall time allocated.

Reply:
That feels about right.

Reply:
But you didn't actually change anything, now the optimiser limit is 110 ms, and you wait for 110 ms. This line needs to be `timeLimitMillis + 10`